ci_framework/plugins/module_utils/networking_mapping/ip_pools.py

```diff
@@ -0,0 +1,129 @@
+import ipaddress
+import typing
+
+
+from ansible_collections.cifmw.general.plugins.module_utils.networking_mapping import (
+    exceptions,
+    networking_definition,
+)
+
+
+class IPPool:
+    def __init__(
+        self,
+        ip_range: networking_definition.HostNetworkRange,
+        reservations: typing.List[
+            typing.Union[ipaddress.IPv4Address, ipaddress.IPv6Address, str]
+        ] = None,
+    ) -> None:
+        self.__index = 0
+        self.__range = ip_range
+        self.__reservations = set()
+        self.__parse_reservations(reservations)
+
+    def add_reservation(
+        self,
+        reservation: typing.Union[ipaddress.IPv4Address, ipaddress.IPv6Address, str],
+    ):
+        reservation_ip = ipaddress.ip_address(reservation)
+        if reservation_ip not in self.__range:
+            raise exceptions.NetworkMappingError(
+                f"Reservation {reservation} is out of range {self.__range}"
+            )
+        self.__reservations.add(reservation_ip)
```

Comment:
Ip reservation would be a typical case where we need to worry about concurrency and add locks, but AFAIU we should not have that scenario while doing a simple mapping, right?

Reply:
The code will run as a module/action plugins, in one single thread, so no idea why we need to think about concurrency. This functionality is used to avoid the pool assigning IPs for instances that the uses has explicitly set, nothing related to the OS, libvirt or whatever low level tool that manages networking.